docs/design/data_safety.md

```diff
@@ -0,0 +1,153 @@
+Data Safety in ICU4X
+====================
+
+At its core, ICU4X is a set of algorithms that map input data to human-ready outputs. Since a core value proposition of ICU4X is the ability to dynamically load data, we can't always assume that our data are valid and pristine. This doc covers how we think about this problem and handle it in ICU4X.
+
+## Resolution
+
+1. _Whereas_ it is rare to be 100% confident about the safety of your data\*; and
+1. _Whereas_ validating data invariants can be expensive; and
+1. _Whereas_ moving validation into the algorithm (post-deserialization) often reduces the overall overhead of validation; and
+1. _Whereas_ it is useful to _be able to_ learn whether code traverses unexpected code paths; and
+1. _Whereas_ end users of ICU4X algorithms are not in a position to reason about invalid data in otherwise-infallible terminal functions; and
+1. _Whereas_ algorithms that panic on malformed data increase the vulnerability space of an application\*\*; and
+1. _Whereas_ "garbage in, garbage out" (GIGO) does not increase the space that malicious actors could leverage\*\*; be it
```

Comment:
Chances are that this isn't entirely true.

It's conceivable that some security property in some application would depend on canonical equivalence of strings (consisting of assigned code points) being computed consistently by different processes / code paths. If the normalization data is corrupt in one process / code path, a canonical equivalence check in one place might not match a canonical equivalence check in another.

It seems that this point is more likely to be true in the case of operations where _locale-dependent_ CLDR data is used for human-readable formatting and less likely to be true in the case of operations where immutable-by-policy locale-independent Unicode Database data is used for querying fundamental characteristics of characters/strings.

Reply:
Yes, but does _GIGO_ increase the vulnerability space?

If malicious actors are able to mutate the UCD data consumed by ICU4X, they can remap character properties as desired, like moving a character from one general category to another. This isn't something that validation could catch, and it's not something that GIGO would enable or prevent.

Reply:
OK. Good point.